plugins/tictactoe/tictactoe.7
.\"
.\" Copyright (c) 2013-2026 David Demelier <markand@malikania.fr>
.\"
.\" Permission to use, copy, modify, and/or distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd @IRCCD_MAN_DATE@
.Dt IRCCD-PLUGIN-TICTACTOE 7
.Os
.\" NAME
.Sh NAME
.Nm tictactoe
.Nd irccd tictactoe plugin
.\" DESCRIPTION
.Sh DESCRIPTION
The
.Nm
plugin let you play tictactoe over IRC.
.Pp
Warning: this plugin is extremely verbose, don't enable it blindly.
.\" INSTALLATION
.Sh INSTALLATION
is distributed with irccd. To enable it add it into your
.Xr irccd.conf 5
configuration file.
.Pp
.Bd -literal -offset indent
plugin "tictactoe"
.Ed
.\" USAGE
.Sh USAGE
Use the special command with the target opponent nickname. Then each player send
a message in the form
.Ar "x y"
where x targets the column and y the row.
.Pp
If one of the players leaves the channel (either by kick or part) the game is
aborted silently. If the player don't play in the delay of five minutes, the
game is aborted to let another game starts.
.Pp
Example when starting a game:
.Bd -literal -offset Ds
markand: !tictactoe francis
irccd:   a b c
irccd: 1 . . .
irccd: 2 . . .
irccd: 3 . . .
irccd: markand, it's your turn
.Ed
.Pp
And then, placing tokens:
.Bd -literal -offset Ds
20:27 < markand> a 1
20:27 < irccd>   a b c
20:27 < irccd> 1 x . .
20:27 < irccd> 2 . . .
20:27 < irccd> 3 . . .
20:27 < irccd> francis, it's your turn
20:27 <@francis> c 1
20:27 < irccd>   a b c
20:27 < irccd> 1 x . o
20:27 < irccd> 2 . . .
20:27 < irccd> 3 . . .
20:27 < irccd> markand, it's your turn
20:27 < markand> a 2
20:27 < irccd>   a b c
20:27 < irccd> 1 x . o
20:27 < irccd> 2 x . .
20:27 < irccd> 3 . . .
20:27 < irccd> francis, it's your turn
20:27 <@francis> c 3
20:27 < irccd>   a b c
20:27 < irccd> 1 x . o
20:27 < irccd> 2 x . .
20:27 < irccd> 3 . . o
20:27 < irccd> markand, it's your turn
20:27 < markand> a 3
20:27 < irccd>   a b c
20:27 < irccd> 1 x . o
20:27 < irccd> 2 x . .
20:27 < irccd> 3 x . o
20:27 < irccd> francis, it's your turn
20:27 < irccd> markand, congratulations, you won!
.Ed
.\" TEMPLATES
.Sh TEMPLATES
The plugin
.Nm
supports the following templates properties:
.Bl -tag -width running
.It Va draw
When the game ended with no winner.
.It Va invalid
The opponent does not exist or is not valid. Keywords:
.Em origin .
.It Va running
The game is already running. Keywords:
.Em origin .
.It Va timeout
The player didn't play within its turn after a certain amount of time.
.It Va turn
Message sent when current player change.
.It Va used
The cell requested is already used.
.It Va win
Game ended with a winner.
.El
.Pp
All templates supports the following standard keywords:
.Em channel , command , nickname , plugin , server .
.\" SEE ALSO
.Sh SEE ALSO
.Xr irccd 1 ,
.Xr irccd-templates 7


CHANGES.md
IRC Client Daemon CHANGES
=========================

irccd 5.0.0 ????-??-??
======================

This is a major release. See MIGRATING.md file for more information.

The key highlight of this release is the integration of libev as the main loop
and coroutines, making the code simpler and more expressive.

Build system
------------

The build system has been changed from CMake to GNU Make, this is because CMake
makes simple things complicated.

C API
-----

- Most of the string based types are dynamically allocated rather than creating
  dozens of arbitrary fixed size limits.
- Functions for allocated types are now coupled as `foo_new` and `foo_free` or
  `foo_init` and `foo_finish` otherwise.
- Types that require string list are now implemented using NULL sentinel based
  lists.

irccd
-----

- The auto-reconnect feature is automatically enabled on a server and is no
  longer an option. As an IRC bot, it make sense that a server should be
  reconnected permanently.
- All deferred functions have been removed.
- The server connection procedure is now implemented using coroutines for a
  more readable code logic.

misc
----

The C and Javascript APIs are now documented using doxygen rather than manual
pages.

irccd 4.1.1 2026-01-07
======================

- Fix minor issue in Irccd.Http API when `curl_multi_fdset` can return `-1` file
  descriptor.

irccd 4.1.0 2025-04-23
======================

- Add `irc_bot_pollable_add` to insert a custom interface into the main irccd
  loop.
- A new option `IRCCD_WITH_HTTP` has been added to accommodate the new curl
  dependency for `Irccd.Http` API.
- The plugin `history` manages two different timestamps for message and
  visibility. This helps showing the last message date of someone who is present
  on a channel but idle.

javascript API
--------------

- Brand new asynchronous `Irccd.Http` API.

plugins
-------

- links: rewritten in Javascript using the brand new `Irccd.Http` API.

irccd 4.0.4 2023-08-03
======================

- Improve tests running on a slow machines.
- Fix a warning on a invalid comparison.

irccd 4.0.3 2023-02-08
======================

- Fix build on NetBSD 9.3.

irccd 4.0.2 2023-01-18
======================

- Fix crash because of undefined variable.

irccd 4.0.1 2022-02-07
======================

- Fix disconnection with multiple servers.
- Fix Javascript `Irccd.Server` constructor function.
- Substitute the irccd version in Javascript plugins.
- Fix `irccdctl plugin-reload` command without argument.

irccd 4.0.0 2022-02-03
======================

This is a major release. See MIGRATING.md file for more information.

The key highlight for this release is the rewrite from C++ to C. The only
runtime dependency required is OpenSSL (if built with SSL support).

irccd
-----

- Irccd keeps track of nicknames in channels by capturing join/part/kick and
  mode changes. It is now more convenient from the plugins to quickly inspect if
  someone is present on a channel.
- It is now possible to change uid/gid of the transport socket file. The file is
  also created with permissions 664.
- A new `paths` command has been added to show default paths.

irccdctl
--------

- Commands `plugin-reload` and `plugin-unload` can be invoked without arguments.
- New `plugin-template` and `plugin-path` command which are synonyms of
  `plugin-config` but for templates and paths respectively.

plugins
-------

- tictactoe: now has a timeout in case of inactivity.

misc
----

- Split irccd-api manual page into individual irccd-api-<module> for a better
  readability.
- New `irccd.conf` and `irccdctl.conf` syntax.

network API
-----------

- Network protocol uses plain text again.
- Transport uses clear UNIX sockets only without passwords.

javascript API
--------------

- Brand new Irccd.Rule API to inspect and manage rules.
- Brand new Irccd.Hook API to inspect and manage hooks.

irccd 3.1.1 2021-01-04
======================

- Synchronize `ping-timeout` option in `[server]` to 1800 seconds by default.
- Enable `auto-reconnect` option in `[server]` by default as specified in the
  manual page.

irccd 3.1.0 2020-07-03
======================

- Added a new hook system. Hooks consist of an alternative approach to plugins
  to extend irccd in any language.

irccd 3.0.3 2019-10-06
======================

- Fix errors in irccdctl.conf example file.
- Add example of password in irccdctl.conf and irccd.conf.

irccd 3.0.2 2019-09-22
======================

- Added *IRCCD_WITH_JS* CMake variable in irccd package.
- Fixed trailing CTCP escape code.
- Fixed invalid parameters in topic event.
- Fixed invalid documentation of Irccd.System.usleep function.
- Fixed invalid system configuration directory.

irccd 3.0.1 2019-09-01
======================

- Fixed an invalid template escape sequence.
- Updated the default configuration files.
- Fix RPATH handling for private libraries like Duktape.

irccd 3.0.0 2019-08-15
======================

This is a major release. See MIGRATING.md file for more information.

irccd
-----

- New sections `[paths]` and `[paths.plugin]` have been added to control
  standard paths for both irccd and plugins.
- Irccd no longer supports uid, gid, pid and daemon features.
- Sections `[identity]` and `[server]` have been merged.
- Local transports support SSL.
- The origin in rule is now first class value.
- New option `ipv4` in `[transport]`.
- New option `ipv4` in `[server]`.
- Section `[format]` is renamed to `[templates]`.
- New commands are available as irccd arguments `info` and `version`.

irccdctl
--------

- New option `ipv4` in `[connect]`.
- New option `-o` in `rule-add`.
- New option `-o` and `-O` in `rule-edit`.

irccd-test
----------

- A brand new `irccd-test` program has been added to tests plugins on the
  command line.

cmake
-----

- CMake no longer create a fake installation directory while building.
- All targets are placed into the `bin` directory while building.

network API
-----------


plugins/tictactoe/tictactoe.js
/*
 * tictactoe.js -- tictactoe game for IRC
 *
 * Copyright (c) 2013-2026 David Demelier <markand@malikania.fr>
 *
 * Permission to use, copy, modify, and/or distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

// Plugin information.
info = {
	author: "David Demelier <markand@malikania.fr>",
	license: "ISC",
	summary: "A tictactoe game for IRC",
	version: "@irccd_VERSION@"
};

// Modules.
var Plugin = Irccd.Plugin;
var Util = Irccd.Util;
var Timer = Irccd.Timer;

// Formats.
Plugin.templates = {
	"draw":         "Nobody won.",
	"invalid":      "#{nickname}, please select a valid opponent.",
	"running":      "#{nickname}, the game is already running.",
	"turn":         "#{nickname}, it's your turn.",
	"used":         "#{nickname}, this square is already used.",
	"win":          "#{nickname}, congratulations, you won!",
	"timeout":      "Aborted due to #{nickname} inactivity."
};

/**
 * Create a game.
 *
 * This function creates a game without any checks.
 *
 * @param server the server object
 * @param channel the channel
 * @param origin the source origin
 * @param target the target nickname
 */
function Game(server, channel, origin, target)
{
	this.server = server;
	this.origin = origin;
	this.target = target;
	this.channel = channel;
	this.players = [ Util.splituser(origin), target ];
	this.player = Math.floor(Math.random() * 2);
	this.grid = [
		[ '.', '.', '.' ],
		[ '.', '.', '.' ],
		[ '.', '.', '.' ]
	];
}

// Pending games requests checking for names listing.
Game.requests = {};

// List of games running.
Game.map = {};

/**
 * Create a unique id.
 *
 * @param server the server object
 * @param channel the channel
 * @return the id
 */
Game.id = function (server, channel)
{
	return channel + "@" + server.toString();
}

/**
 * Get a running game or undefined.
 *
 * @param server the server object
 * @param channel the channel
 * @return the object or undefined if not running
 */
Game.find = function (server, channel)
{
	return Game.map[Game.id(server, channel)];
}

/**
 * Populate a set of keywords.
 *
 * @param server the server object
 * @param origin the originator
 * @param channel the channel
 * @return an object of predefined keywords
 */
Game.keywords = function (server, channel, origin)
{
	var kw = {
		channel: channel,
		command: server.info().prefix + Plugin.info().name,
		plugin: Plugin.info().name,
		server: server.info().name
	};

	if (origin) {
		kw.origin = origin;
		kw.nickname = Util.splituser(origin);
	}

	return kw;
}

/**
 * Tells if a game is pending or running.
 *
 * @param server the server object
 * @param channel the channel
 * @return true if any
 */
Game.exists = function (server, channel)
{
	var id = Game.id(server, channel);

	return Game.requests[id] || Game.map[id];
}

/**
 * Delete a game from the registry.
 *
 * @param server the server object
 * @param channel the channel
 */
Game.remove = function (server, channel)
{
	delete Game.map[Game.id(server, channel)];
}

/**
 * Erase games when some players leave channels.
 *
 * @param server the server object
 * @param origin the originator
 * @param channel the channel
 */
Game.clear = function (server, user, channel)
{
	var nickname = Util.splituser(user);
	var game = Game.find(server, channel);

	if (game && (game.players[0] === nickname || game.players[1] === nickname))
		Game.remove(server, channel);
}

/**
 * Check if the target is valid.
 *
 * @param server the server object
 * @param channel the channel string
 * @param nickname the nickname who requested the game
 * @param target the opponent
 * @return true if target is valid
 */
Game.isValid = function (server, channel, nickname, target)
{
	if (target === "" || target === nickname || target === server.info().nickname)
		return false;

	var channels = server.info().channels;
	var ch;

	for (var i = 0; i < channels.length; ++i) {
		if (channels[i].name === channel) {
			ch = channels[i];
			break;
		}
	}

	for (var i = 0; i < ch.users.length; ++i)
		if (ch.users[i].nickname === target)
			return true;

	return false;
}

/**
 * Function called when a timeout occured.
 *
 * @param game the game to destroy
 */
Game.timeout = function (game)
{
	var kw = Game.keywords(game.server, game.channel);

	kw.nickname = game.players[game.player];
	game.server.message(game.channel, Util.format(Plugin.templates.timeout, kw));
	Game.remove(game.server, game.channel);
}

/**
 * Show the game grid and the next player line.
 */
Game.prototype.show = function ()
{
	var kw = Game.keywords(this.server, this.channel);
	var self = this;

	// nickname is the current player.
	kw.nickname = this.players[this.player];

	this.server.message(this.channel, "  a b c");
	this.server.message(this.channel, "1 " + this.grid[0].join(" "));
	this.server.message(this.channel, "2 " + this.grid[1].join(" "));
	this.server.message(this.channel, "3 " + this.grid[2].join(" "));

	if (this.hasWinner())
		this.server.message(this.channel, Util.format(Plugin.templates.win, kw));
	else if (this.hasDraw())
		this.server.message(this.channel, Util.format(Plugin.templates.draw, kw));
	else
		this.server.message(this.channel, Util.format(Plugin.templates.turn, kw));

	// Create a timer in case of inactivity (5 minutes).
	this.timer = new Irccd.Timer(Irccd.Timer.Single, 300000, function () {
		Game.timeout(self);
	});
	this.timer.start();
}

/**
 * Tells if it's the nickname's turn.
 *
 * @param nickname the nickname to check
 * @return true if nickname is allowed to play
 */
Game.prototype.isTurn = function (nickname)
{
	return this.players[this.player] == nickname;
}

/**
 * Place the column and row as the current player.
 *
 * @param column the column (a, b or c)
 * @param row the row (1, 2 or 3)
 */
Game.prototype.place = function (column, row, origin)
{
	var columns = { a: 0, b: 1, c: 2 };
	var rows = { 1: 0, 2: 1, 3: 2 };

	column = columns[column];
	row = rows[row];

	var kw = Game.keywords(this.server, this.channel, origin);

	if (this.grid[row][column] !== '.') {
		this.server.message(this.channel, Util.format(Plugin.templates.used, kw));
		return false;
	}

	this.timer.stop();
	this.grid[row][column] = this.player === 0 ? 'x' : 'o';

	// Do not change if game is finished.
	if (!this.hasWinner() && !this.hasDraw())
		this.player = this.player === 0 ? 1 : 0;

	return true;
}

/**
 * Check if there is a winner.
 *
 * @return true if there is a winner
 */
Game.prototype.hasWinner = function ()
{
	var lines = [
		[ [ 0, 0 ], [ 0, 1 ], [ 0, 2 ] ],
		[ [ 1, 0 ], [ 1, 1 ], [ 1, 2 ] ],
		[ [ 2, 0 ], [ 2, 1 ], [ 2, 2 ] ],
		[ [ 0, 0 ], [ 1, 0 ], [ 2, 0 ] ],
		[ [ 0, 1 ], [ 1, 1 ], [ 2, 1 ] ],
		[ [ 0, 2 ], [ 1, 2 ], [ 2, 2 ] ],
		[ [ 0, 0 ], [ 1, 1 ], [ 2, 2 ] ],
		[ [ 0, 2 ], [ 1, 1 ], [ 2, 0 ] ]
	];

	for (var i = 0; i < lines.length; ++i) {
		var p1 = lines[i][0];
		var p2 = lines[i][1];
		var p3 = lines[i][2];

		var result = this.grid[p1[0]][p1[1]] === this.grid[p2[0]][p2[1]] &&
		             this.grid[p2[0]][p2[1]] === this.grid[p3[0]][p3[1]] &&
		             this.grid[p3[0]][p3[1]] !== '.';

		if (result)
			return true;
	}
}

/**
 * Check if there is draw game.
 *
 * @return true if game is draw
 */
Game.prototype.hasDraw = function ()
{
	for (var r = 0; r < 3; ++r)
		for (var c = 0; c < 3; ++c)
			if (this.grid[r][c] === '.')
				return false;

	return true;
}

function onCommand(server, origin, channel, message)
{
	channel = channel.toLowerCase();

	var target = message.trim();
	var nickname = Util.splituser(origin);

	if (Game.exists(server, channel))
		server.message(channel, Util.format(Plugin.templates.running, Game.keywords(server, channel, origin)));
	else if (!Game.isValid(server, channel, nickname, target))
		server.message(channel, Util.format(Plugin.templates.invalid, Game.keywords(server, channel, origin)));
	else {
		var game = new Game(server, channel, origin, target);

		Game.map[Game.id(server, channel)] = game;
		game.show();
	}
}

function onMessage(server, origin, channel, message)
{
	channel = channel.toLowerCase();

	var nickname = Util.splituser(origin);
	var game = Game.find(server, channel);

	if (!game || !game.isTurn(nickname))
		return;

	var match = /^([abc]) ?([123])$/.exec(message.trim());

	if (!match)
		return;

	if (game.place(match[1], match[2], origin))
		game.show();
	if (game.hasWinner() || game.hasDraw())
		Game.remove(server, channel);
}

function onDisconnect(server)
{
	for (var key in Game.map)
		if (key.endsWith(server.toString()))
			delete Game.map[key];
}

function onKick(server, origin, channel, target)
{
	Game.clear(server, target, channel.toLowerCase());
}

function onPart(server, origin, channel)
{
	Game.clear(server, origin, channel.toLowerCase());
}
